private/buf/bufformat/formatter.go

```diff
@@ -1443,12 +1443,44 @@ func (f *formatter) writeSpecialFloatLiteral(specialFloatLiteralNode *ast.Specia
 
 // writeStringLiteral writes a string literal value (e.g. "foo").
 func (f *formatter) writeStringLiteral(stringLiteralNode *ast.StringLiteralNode) {
-	f.WriteString(fmt.Sprintf("%q", stringLiteralNode.Val))
+	f.writeStringWithStyle(stringLiteralNode.Val)
 }
 
 // writeStringValue writes a string value (e.g. "foo").
 func (f *formatter) writeStringValue(stringValueNode ast.StringValueNode) {
-	f.WriteString(fmt.Sprintf("%q", stringValueNode.AsString()))
+	f.writeStringWithStyle(stringValueNode.AsString())
+}
+
+// writeStringWithStyle writes a string value with surrounding single quotes or double quotes
+// based on the string value's content. The formatter prefers double quotes, but uses single
+// quotes if the content contains any (") or (\") literals and not any (') or (\') literals.
+//
+// For example,
+//
+//  1. f"o"'o' -> "f\"o\"'o'"      (") and (') are used - surround with (")
+//  2. f"oo" -> 'f"oo"'            (") is used - surround with (')
+//  3. f'oo' -> "f'oo'"            (') is used - surround with (")
+//  4. f\"o\"\'o\' -> "f\"o\"'o'"  (\") and (\') are used - surround with (")
+//  5. f\"o\"o -> 'f"o"o'          (\") is used - surround with (')
```

Comment:
Won't this be `f\"o\"o` and not `f"o"o` as written?

Reply:
Not with what I'm going for here - if the user escaped `"` characters, but the string didn't include any `'`, then we can safely surround it with `'` (as shown).

Reply:
Where do we unquote the input though?

https://go.dev/play/p/sLN2nHYBHs5

Reply:
My mistake, I was interpreting this example as if it were surrounding in `"` (like [this](https://go.dev/play/p/Euv-btoqzzk)). But now I realize what you're calling out - I'll take a closer look.

Reply:
Ok, I saw the problem - we don't actually need to care about the explicit escape literals at all in this case. I simplified it down so that it's more clear.